File: compatibility-tests/sdk-test-java/src/test/java/com/floci/test/PipesTest.java
package com.floci.test;

import org.junit.jupiter.api.*;
import software.amazon.awssdk.services.pipes.PipesClient;
import software.amazon.awssdk.services.pipes.model.*;
import software.amazon.awssdk.services.sqs.SqsClient;
import software.amazon.awssdk.services.sqs.model.*;

import static org.assertj.core.api.Assertions.*;

@DisplayName("EventBridge Pipes")
@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
class PipesTest {

    private static final String ACCOUNT_ID = "000000000000";
    private static final String REGION = "us-east-1";
    private static final String ROLE_ARN = "arn:aws:iam::" + ACCOUNT_ID + ":role/pipe-role";

    private static PipesClient pipes;
    private static SqsClient sqs;
    private static String pipeName;
    private static String srcQueue;
    private static String tgtQueue;
    private static String srcQueueUrl;
    private static String tgtQueueUrl;

    private static String sqsArn(String queueName) {
        return "arn:aws:sqs:" + REGION + ":" + ACCOUNT_ID + ":" + queueName;
    }

    @BeforeAll
    static void setup() {
        pipes = TestFixtures.pipesClient();
        sqs = TestFixtures.sqsClient();
        pipeName = TestFixtures.uniqueName("pipe");
        srcQueue = TestFixtures.uniqueName("pipe-src");
        tgtQueue = TestFixtures.uniqueName("pipe-tgt");

        srcQueueUrl = sqs.createQueue(CreateQueueRequest.builder()
                .queueName(srcQueue).build()).queueUrl();
        tgtQueueUrl = sqs.createQueue(CreateQueueRequest.builder()
                .queueName(tgtQueue).build()).queueUrl();
    }

    @AfterAll
    static void cleanup() {
        if (pipes != null) {
            try { pipes.deletePipe(DeletePipeRequest.builder().name(pipeName).build()); } catch (Exception ignored) {}
            pipes.close();
        }
        if (sqs != null) {
            try { sqs.deleteQueue(DeleteQueueRequest.builder().queueUrl(srcQueueUrl).build()); } catch (Exception ignored) {}
            try { sqs.deleteQueue(DeleteQueueRequest.builder().queueUrl(tgtQueueUrl).build()); } catch (Exception ignored) {}
            sqs.close();
        }
    }

    @Test
    @Order(1)
    void createPipe() {
        CreatePipeResponse response = pipes.createPipe(CreatePipeRequest.builder()
                .name(pipeName)
                .source(sqsArn(srcQueue))
                .target(sqsArn(tgtQueue))
                .roleArn(ROLE_ARN)
                .desiredState(RequestedPipeState.STOPPED)
                .build());

        assertThat(response.currentState()).isEqualTo(PipeState.STOPPED);
        assertThat(response.arn()).contains(pipeName);
    }

    @Test
    @Order(2)
    void describePipe() {
        DescribePipeResponse response = pipes.describePipe(DescribePipeRequest.builder()
                .name(pipeName).build());

        assertThat(response.name()).isEqualTo(pipeName);
        assertThat(response.source()).isEqualTo(sqsArn(srcQueue));
        assertThat(response.target()).isEqualTo(sqsArn(tgtQueue));
        assertThat(response.currentState()).isEqualTo(PipeState.STOPPED);
    }

    @Test
    @Order(3)
    void listPipes() {
        ListPipesResponse response = pipes.listPipes(ListPipesRequest.builder().build());

        assertThat(response.pipes())
                .anyMatch(p -> pipeName.equals(p.name()));
    }

    @Test
    @Order(4)
    void updatePipe() {
        pipes.updatePipe(UpdatePipeRequest.builder()
                .name(pipeName)
                .roleArn(ROLE_ARN)
                .description("updated via SDK")
                .desiredState(RequestedPipeState.STOPPED)
                .build());

        DescribePipeResponse response = pipes.describePipe(DescribePipeRequest.builder()
                .name(pipeName).build());

        assertThat(response.description()).isEqualTo("updated via SDK");
    }

    @Test
    @Order(5)
    void startAndStopPipe() {
        StartPipeResponse startResponse = pipes.startPipe(StartPipeRequest.builder()
                .name(pipeName).build());
        assertThat(startResponse.currentState()).isEqualTo(PipeState.RUNNING);

        StopPipeResponse stopResponse = pipes.stopPipe(StopPipeRequest.builder()
                .name(pipeName).build());
        assertThat(stopResponse.currentState()).isEqualTo(PipeState.STOPPED);
    }

    @Test
    @Order(6)
    void deletePipe() {
        pipes.deletePipe(DeletePipeRequest.builder().name(pipeName).build());

        assertThatThrownBy(() -> pipes.describePipe(DescribePipeRequest.builder()
                .name(pipeName).build()))
                .isInstanceOf(NotFoundException.class);
    }

    @Test
    @Order(7)
    void describeNonExistentPipe() {
        assertThatThrownBy(() -> pipes.describePipe(DescribePipeRequest.builder()
                .name("nonexistent-pipe").build()))
                .isInstanceOf(NotFoundException.class);
    }

    @Test
    @Order(10)
    void sqsToSqsForwarding() throws InterruptedException {
        String fwdPipeName = TestFixtures.uniqueName("pipe-fwd");
        String fwdSrc = TestFixtures.uniqueName("pipe-fwd-src");
        String fwdTgt = TestFixtures.uniqueName("pipe-fwd-tgt");
        String fwdSrcUrl = sqs.createQueue(CreateQueueRequest.builder()
                .queueName(fwdSrc).build()).queueUrl();
        String fwdTgtUrl = sqs.createQueue(CreateQueueRequest.builder()
                .queueName(fwdTgt).build()).queueUrl();

        try {
            pipes.createPipe(CreatePipeRequest.builder()
                    .name(fwdPipeName)
                    .source(sqsArn(fwdSrc))
                    .target(sqsArn(fwdTgt))
                    .roleArn(ROLE_ARN)
                    .desiredState(RequestedPipeState.RUNNING)
                    .build());

            sqs.sendMessage(SendMessageRequest.builder()
                    .queueUrl(fwdSrcUrl)
                    .messageBody("hello from pipes")
                    .build());

            boolean found = false;
            for (int i = 0; i < 15; i++) {
                Thread.sleep(1000);
                ReceiveMessageResponse recv = sqs.receiveMessage(ReceiveMessageRequest.builder()
                        .queueUrl(fwdTgtUrl)
                        .maxNumberOfMessages(1)
                        .waitTimeSeconds(1)
                        .build());
                if (!recv.messages().isEmpty()
                        && recv.messages().get(0).body().contains("hello from pipes")) {
                    found = true;
                    break;
                }
            }
            assertThat(found).as("target queue should receive forwarded message").isTrue();
        } finally {
            try { pipes.deletePipe(DeletePipeRequest.builder().name(fwdPipeName).build()); } catch (Exception ignored) {}
            try { sqs.deleteQueue(DeleteQueueRequest.builder().queueUrl(fwdSrcUrl).build()); } catch (Exception ignored) {}
            try { sqs.deleteQueue(DeleteQueueRequest.builder().queueUrl(fwdTgtUrl).build()); } catch (Exception ignored) {}
        }
    }

    @Test
    @Order(11)
    void filterCriteriaFiltersMessages() throws InterruptedException {
        String filterPipeName = TestFixtures.uniqueName("pipe-filter");
        String filterSrc = TestFixtures.uniqueName("pipe-filter-src");
        String filterTgt = TestFixtures.uniqueName("pipe-filter-tgt");
        String filterSrcUrl = sqs.createQueue(CreateQueueRequest.builder()
                .queueName(filterSrc).build()).queueUrl();
        String filterTgtUrl = sqs.createQueue(CreateQueueRequest.builder()
                .queueName(filterTgt).build()).queueUrl();

        try {
            pipes.createPipe(CreatePipeRequest.builder()
                    .name(filterPipeName)
                    .source(sqsArn(filterSrc))
                    .target(sqsArn(filterTgt))
                    .roleArn(ROLE_ARN)
                    .desiredState(RequestedPipeState.RUNNING)
                    .sourceParameters(PipeSourceParameters.builder()
                            .filterCriteria(FilterCriteria.builder()
                                    .filters(Filter.builder()
                                            .pattern("{\"body\": {\"status\": [\"active\"]}}")
                                            .build())
                                    .build())
                            .build())
                    .build());

            sqs.sendMessage(SendMessageRequest.builder()
                    .queueUrl(filterSrcUrl)
                    .messageBody("{\"status\": \"active\", \"id\": \"match-1\"}")
                    .build());
            sqs.sendMessage(SendMessageRequest.builder()
                    .queueUrl(filterSrcUrl)
                    .messageBody("{\"status\": \"inactive\", \"id\": \"no-match\"}")
                    .build());

            boolean found = false;
            for (int i = 0; i < 15; i++) {
                Thread.sleep(1000);
                ReceiveMessageResponse recv = sqs.receiveMessage(ReceiveMessageRequest.builder()
                        .queueUrl(filterTgtUrl)
                        .maxNumberOfMessages(10)
                        .waitTimeSeconds(1)
                        .build());
                if (recv.messages().stream().anyMatch(m -> m.body().contains("match-1"))) {
                    assertThat(recv.messages().stream().noneMatch(m -> m.body().contains("no-match")))
                            .as("non-matching message should not be forwarded").isTrue();
                    found = true;
                    break;
                }
            }
            assertThat(found).as("target queue should receive matching message").isTrue();

            GetQueueAttributesResponse attrs = sqs.getQueueAttributes(GetQueueAttributesRequest.builder()
                    .queueUrl(filterSrcUrl)
                    .attributeNames(QueueAttributeName.APPROXIMATE_NUMBER_OF_MESSAGES)
                    .build());
            assertThat(attrs.attributes().get(QueueAttributeName.APPROXIMATE_NUMBER_OF_MESSAGES)).isEqualTo("0");
        } finally {
            try { pipes.deletePipe(DeletePipeRequest.builder().name(filterPipeName).build()); } catch (Exception ignored) {}
            try { sqs.deleteQueue(DeleteQueueRequest.builder().queueUrl(filterSrcUrl).build()); } catch (Exception ignored) {}
            try { sqs.deleteQueue(DeleteQueueRequest.builder().queueUrl(filterTgtUrl).build()); } catch (Exception ignored) {}
        }
    }

    @Test
    @Order(12)
    void batchSizeInSourceParameters() throws InterruptedException {
        String batchPipeName = TestFixtures.uniqueName("pipe-batch");
        String batchSrc = TestFixtures.uniqueName("pipe-batch-src");
        String batchTgt = TestFixtures.uniqueName("pipe-batch-tgt");
        String batchSrcUrl = sqs.createQueue(CreateQueueRequest.builder()
                .queueName(batchSrc).build()).queueUrl();
        String batchTgtUrl = sqs.createQueue(CreateQueueRequest.builder()
                .queueName(batchTgt).build()).queueUrl();

        try {
            pipes.createPipe(CreatePipeRequest.builder()
                    .name(batchPipeName)
                    .source(sqsArn(batchSrc))
                    .target(sqsArn(batchTgt))
                    .roleArn(ROLE_ARN)
                    .desiredState(RequestedPipeState.RUNNING)
                    .sourceParameters(PipeSourceParameters.builder()
                            .sqsQueueParameters(PipeSourceSqsQueueParameters.builder()
                                    .batchSize(1)
                                    .build())
                            .build())
                    .build());

            for (int i = 1; i <= 3; i++) {
                sqs.sendMessage(SendMessageRequest.builder()
                        .queueUrl(batchSrcUrl)
                        .messageBody("batch-msg-" + i)
                        .build());
            }

            java.util.Set<String> foundMessages = new java.util.HashSet<>();
            for (int i = 0; i < 20 && foundMessages.size() < 3; i++) {
                Thread.sleep(1000);
                ReceiveMessageResponse recv = sqs.receiveMessage(ReceiveMessageRequest.builder()
                        .queueUrl(batchTgtUrl)
                        .maxNumberOfMessages(10)
                        .waitTimeSeconds(1)
                        .build());
                for (Message msg : recv.messages()) {
                    for (int j = 1; j <= 3; j++) {
                        if (msg.body().contains("batch-msg-" + j)) {
                            foundMessages.add("batch-msg-" + j);
                        }
                    }
                }
            }
            assertThat(foundMessages).hasSize(3);
        } finally {
            try { pipes.deletePipe(DeletePipeRequest.builder().name(batchPipeName).build()); } catch (Exception ignored) {}
            try { sqs.deleteQueue(DeleteQueueRequest.builder().queueUrl(batchSrcUrl).build()); } catch (Exception ignored) {}
            try { sqs.deleteQueue(DeleteQueueRequest.builder().queueUrl(batchTgtUrl).build()); } catch (Exception ignored) {}
        }
    }

    @Test
    @Order(13)
    void stoppedPipeDoesNotForward() throws InterruptedException {
        String nfPipeName = TestFixtures.uniqueName("pipe-nofwd");
        String nfSrc = TestFixtures.uniqueName("pipe-nofwd-src");
        String nfTgt = TestFixtures.uniqueName("pipe-nofwd-tgt");
        String nfSrcUrl = sqs.createQueue(CreateQueueRequest.builder()
                .queueName(nfSrc).build()).queueUrl();
        String nfTgtUrl = sqs.createQueue(CreateQueueRequest.builder()
                .queueName(nfTgt).build()).queueUrl();

        try {
            pipes.createPipe(CreatePipeRequest.builder()
                    .name(nfPipeName)
                    .source(sqsArn(nfSrc))
                    .target(sqsArn(nfTgt))
                    .roleArn(ROLE_ARN)
                    .desiredState(RequestedPipeState.STOPPED)
                    .build());

            sqs.sendMessage(SendMessageRequest.builder()
                    .queueUrl(nfSrcUrl)
                    .messageBody("should not forward")
                    .build());

            Thread.sleep(3000);

            ReceiveMessageResponse recv = sqs.receiveMessage(ReceiveMessageRequest.builder()
                    .queueUrl(nfTgtUrl)
                    .maxNumberOfMessages(1)
                    .waitTimeSeconds(1)
                    .build());
            assertThat(recv.messages()).isEmpty();
        } finally {
            try { pipes.deletePipe(DeletePipeRequest.builder().name(nfPipeName).build()); } catch (Exception ignored) {}
            try { sqs.deleteQueue(DeleteQueueRequest.builder().queueUrl(nfSrcUrl).build()); } catch (Exception ignored) {}
            try { sqs.deleteQueue(DeleteQueueRequest.builder().queueUrl(nfTgtUrl).build()); } catch (Exception ignored) {}
        }
    }
}

File: compatibility-tests/sdk-test-rust/tests/pipes_test.rs
mod common;

const ACCOUNT_ID: &str = "000000000000";
const REGION: &str = "us-east-1";
const ROLE_ARN: &str = "arn:aws:iam::000000000000:role/pipe-role";

fn sqs_arn(queue_name: &str) -> String {
    format!("arn:aws:sqs:{}:{}:{}", REGION, ACCOUNT_ID, queue_name)
}

#[tokio::test(flavor = "multi_thread")]
async fn test_pipes_create_pipe() {
    let pipes = common::pipes_client().await;
    let sqs = common::sqs_client().await;
    let src_queue = "rust-pipe-src-create";
    let tgt_queue = "rust-pipe-tgt-create";
    let pipe_name = "rust-pipe-create";

    sqs.create_queue().queue_name(src_queue).send().await.expect("create src queue");
    sqs.create_queue().queue_name(tgt_queue).send().await.expect("create tgt queue");

    let _guard = common::CleanupGuard::new({
        let pipes = pipes.clone();
        let sqs = sqs.clone();
        async move {
            let _ = pipes.delete_pipe().name(pipe_name).send().await;
            cleanup_queue(&sqs, src_queue).await;
            cleanup_queue(&sqs, tgt_queue).await;
        }
    });

    let result = pipes
        .create_pipe()
        .name(pipe_name)
        .source(sqs_arn(src_queue))
        .target(sqs_arn(tgt_queue))
        .role_arn(ROLE_ARN)
        .desired_state(aws_sdk_pipes::types::RequestedPipeState::Stopped)
        .send()
        .await;
    assert!(result.is_ok(), "CreatePipe failed: {:?}", result.err());

    let output = result.unwrap();
    assert_eq!(
        output.current_state(),
        Some(&aws_sdk_pipes::types::PipeState::Stopped)
    );
    assert!(output.arn().unwrap_or("").contains(pipe_name));
}

#[tokio::test(flavor = "multi_thread")]
async fn test_pipes_describe_pipe() {
    let pipes = common::pipes_client().await;
    let sqs = common::sqs_client().await;
    let src_queue = "rust-pipe-src-describe";
    let tgt_queue = "rust-pipe-tgt-describe";
    let pipe_name = "rust-pipe-describe";

    sqs.create_queue().queue_name(src_queue).send().await.expect("create src queue");
    sqs.create_queue().queue_name(tgt_queue).send().await.expect("create tgt queue");

    let _guard = common::CleanupGuard::new({
        let pipes = pipes.clone();
        let sqs = sqs.clone();
        async move {
            let _ = pipes.delete_pipe().name(pipe_name).send().await;
            cleanup_queue(&sqs, src_queue).await;
            cleanup_queue(&sqs, tgt_queue).await;
        }
    });

    pipes
        .create_pipe()
        .name(pipe_name)
        .source(sqs_arn(src_queue))
        .target(sqs_arn(tgt_queue))
        .role_arn(ROLE_ARN)
        .desired_state(aws_sdk_pipes::types::RequestedPipeState::Stopped)
        .send()
        .await
        .expect("create pipe");

    let result = pipes.describe_pipe().name(pipe_name).send().await;
    assert!(result.is_ok(), "DescribePipe failed: {:?}", result.err());

    let output = result.unwrap();
    assert_eq!(output.name().unwrap_or(""), pipe_name);
    assert_eq!(output.source().unwrap_or(""), sqs_arn(src_queue));
    assert_eq!(output.target().unwrap_or(""), sqs_arn(tgt_queue));
    assert_eq!(
        output.current_state(),
        Some(&aws_sdk_pipes::types::PipeState::Stopped)
    );
}

#[tokio::test(flavor = "multi_thread")]
async fn test_pipes_list_pipes() {
    let pipes = common::pipes_client().await;
    let sqs = common::sqs_client().await;
    let src_queue = "rust-pipe-src-list";
    let tgt_queue = "rust-pipe-tgt-list";
    let pipe_name = "rust-pipe-list";

    sqs.create_queue().queue_name(src_queue).send().await.expect("create src queue");
    sqs.create_queue().queue_name(tgt_queue).send().await.expect("create tgt queue");

    let _guard = common::CleanupGuard::new({
        let pipes = pipes.clone();
        let sqs = sqs.clone();
        async move {
            let _ = pipes.delete_pipe().name(pipe_name).send().await;
            cleanup_queue(&sqs, src_queue).await;
            cleanup_queue(&sqs, tgt_queue).await;
        }
    });

    pipes
        .create_pipe()
        .name(pipe_name)
        .source(sqs_arn(src_queue))
        .target(sqs_arn(tgt_queue))
        .role_arn(ROLE_ARN)
        .desired_state(aws_sdk_pipes::types::RequestedPipeState::Stopped)
        .send()
        .await
        .expect("create pipe");

    let result = pipes.list_pipes().send().await;
    assert!(result.is_ok(), "ListPipes failed: {:?}", result.err());

    let found = result
        .unwrap()
        .pipes()
        .iter()
        .any(|p| p.name().unwrap_or("") == pipe_name);
    assert!(found, "pipe should appear in list");
}

#[tokio::test(flavor = "multi_thread")]
async fn test_pipes_update_pipe() {
    let pipes = common::pipes_client().await;
    let sqs = common::sqs_client().await;
    let src_queue = "rust-pipe-src-update";
    let tgt_queue = "rust-pipe-tgt-update";
    let pipe_name = "rust-pipe-update";

    sqs.create_queue().queue_name(src_queue).send().await.expect("create src queue");
    sqs.create_queue().queue_name(tgt_queue).send().await.expect("create tgt queue");

    let _guard = common::CleanupGuard::new({
        let pipes = pipes.clone();
        let sqs = sqs.clone();
        async move {
            let _ = pipes.delete_pipe().name(pipe_name).send().await;
            cleanup_queue(&sqs, src_queue).await;
            cleanup_queue(&sqs, tgt_queue).await;
        }
    });

    pipes
        .create_pipe()
        .name(pipe_name)
        .source(sqs_arn(src_queue))
        .target(sqs_arn(tgt_queue))
        .role_arn(ROLE_ARN)
        .desired_state(aws_sdk_pipes::types::RequestedPipeState::Stopped)
        .send()
        .await
        .expect("create pipe");

    pipes
        .update_pipe()
        .name(pipe_name)
        .role_arn(ROLE_ARN)
        .description("updated via SDK")
        .desired_state(aws_sdk_pipes::types::RequestedPipeState::Stopped)
        .send()
        .await
        .expect("update pipe");

    let result = pipes.describe_pipe().name(pipe_name).send().await.expect("describe pipe");
    assert_eq!(result.description().unwrap_or(""), "updated via SDK");
}

#[tokio::test(flavor = "multi_thread")]
async fn test_pipes_delete_pipe() {
    let pipes = common::pipes_client().await;
    let sqs = common::sqs_client().await;
    let src_queue = "rust-pipe-src-del";
    let tgt_queue = "rust-pipe-tgt-del";
    let pipe_name = "rust-pipe-del";

    sqs.create_queue().queue_name(src_queue).send().await.expect("create src queue");
    sqs.create_queue().queue_name(tgt_queue).send().await.expect("create tgt queue");

    let _guard = common::CleanupGuard::new({
        let sqs = sqs.clone();
        async move {
            cleanup_queue(&sqs, src_queue).await;
            cleanup_queue(&sqs, tgt_queue).await;
        }
    });

    pipes
        .create_pipe()
        .name(pipe_name)
        .source(sqs_arn(src_queue))
        .target(sqs_arn(tgt_queue))
        .role_arn(ROLE_ARN)
        .desired_state(aws_sdk_pipes::types::RequestedPipeState::Stopped)
        .send()
        .await
        .expect("create pipe");

    let result = pipes.delete_pipe().name(pipe_name).send().await;
    assert!(result.is_ok(), "DeletePipe failed: {:?}", result.err());

    let describe = pipes.describe_pipe().name(pipe_name).send().await;
    assert!(describe.is_err(), "DescribePipe should fail after deletion");
}

#[tokio::test(flavor = "multi_thread")]
async fn test_pipes_describe_nonexistent() {
    let pipes = common::pipes_client().await;
    let result = pipes.describe_pipe().name("nonexistent-pipe").send().await;
    assert!(result.is_err());
}

#[tokio::test(flavor = "multi_thread")]
async fn test_pipes_start_and_stop() {
    let pipes = common::pipes_client().await;
    let sqs = common::sqs_client().await;
    let src_queue = "rust-pipe-src-startstop";
    let tgt_queue = "rust-pipe-tgt-startstop";
    let pipe_name = "rust-pipe-startstop";

    sqs.create_queue().queue_name(src_queue).send().await.expect("create src queue");
    sqs.create_queue().queue_name(tgt_queue).send().await.expect("create tgt queue");

    let _guard = common::CleanupGuard::new({
        let pipes = pipes.clone();
        let sqs = sqs.clone();
        async move {
            let _ = pipes.delete_pipe().name(pipe_name).send().await;
            cleanup_queue(&sqs, src_queue).await;
            cleanup_queue(&sqs, tgt_queue).await;
        }
    });

    pipes
        .create_pipe()
        .name(pipe_name)
        .source(sqs_arn(src_queue))
        .target(sqs_arn(tgt_queue))
        .role_arn(ROLE_ARN)
        .desired_state(aws_sdk_pipes::types::RequestedPipeState::Stopped)
        .send()
        .await
        .expect("create pipe");

    let start = pipes.start_pipe().name(pipe_name).send().await;
    assert!(start.is_ok(), "StartPipe failed: {:?}", start.err());
    assert_eq!(
        start.unwrap().current_state(),
        Some(&aws_sdk_pipes::types::PipeState::Running)
    );

    let stop = pipes.stop_pipe().name(pipe_name).send().await;
    assert!(stop.is_ok(), "StopPipe failed: {:?}", stop.err());
    assert_eq!(
        stop.unwrap().current_state(),
        Some(&aws_sdk_pipes::types::PipeState::Stopped)
    );
}

#[tokio::test(flavor = "multi_thread")]
async fn test_pipes_sqs_to_sqs_forwarding() {
    let pipes = common::pipes_client().await;
    let sqs = common::sqs_client().await;
    let src_queue = "rust-pipe-src-fwd";
    let tgt_queue = "rust-pipe-tgt-fwd";
    let pipe_name = "rust-pipe-fwd";

    let src_resp = sqs.create_queue().queue_name(src_queue).send().await.expect("create src queue");
    let src_url = src_resp.queue_url().unwrap_or("").to_string();
    let tgt_resp = sqs.create_queue().queue_name(tgt_queue).send().await.expect("create tgt queue");
    let tgt_url = tgt_resp.queue_url().unwrap_or("").to_string();

    let _guard = common::CleanupGuard::new({
        let pipes = pipes.clone();
        let sqs = sqs.clone();
        let src_url = src_url.clone();
        let tgt_url = tgt_url.clone();
        async move {
            let _ = pipes.delete_pipe().name(pipe_name).send().await;
            let _ = sqs.delete_queue().queue_url(&src_url).send().await;
            let _ = sqs.delete_queue().queue_url(&tgt_url).send().await;
        }
    });

    pipes
        .create_pipe()
        .name(pipe_name)
        .source(sqs_arn(src_queue))
        .target(sqs_arn(tgt_queue))
        .role_arn(ROLE_ARN)
        .desired_state(aws_sdk_pipes::types::RequestedPipeState::Running)
        .send()
        .await
        .expect("create pipe");

    sqs.send_message()
        .queue_url(&src_url)
        .message_body("hello from pipes")
        .send()
        .await
        .expect("send message");

    let mut found = false;
    for _ in 0..15 {
        tokio::time::sleep(std::time::Duration::from_secs(1)).await;
        let recv = sqs
            .receive_message()
            .queue_url(&tgt_url)
            .max_number_of_messages(1)
            .wait_time_seconds(1)
            .send()
            .await;
        if let Ok(r) = recv {
            if let Some(msgs) = r.messages() {
                if !msgs.is_empty() && msgs[0].body().unwrap_or("").contains("hello from pipes") {
                    found = true;
                    break;
                }
            }
        }
    }
    assert!(found, "target queue should receive forwarded message");
}

#[tokio::test(flavor = "multi_thread")]
async fn test_pipes_filter_criteria() {
    let pipes = common::pipes_client().await;
    let sqs = common::sqs_client().await;
    let src_queue = "rust-pipe-src-filter";
    let tgt_queue = "rust-pipe-tgt-filter";
    let pipe_name = "rust-pipe-filter";

    let src_resp = sqs.create_queue().queue_name(src_queue).send().await.expect("create src queue");
    let src_url = src_resp.queue_url().unwrap_or("").to_string();
    let tgt_resp = sqs.create_queue().queue_name(tgt_queue).send().await.expect("create tgt queue");
    let tgt_url = tgt_resp.queue_url().unwrap_or("").to_string();

    let _guard = common::CleanupGuard::new({
        let pipes = pipes.clone();
        let sqs = sqs.clone();
        let src_url = src_url.clone();
        let tgt_url = tgt_url.clone();
        async move {
            let _ = pipes.delete_pipe().name(pipe_name).send().await;
            let _ = sqs.delete_queue().queue_url(&src_url).send().await;
            let _ = sqs.delete_queue().queue_url(&tgt_url).send().await;
        }
    });

    pipes
        .create_pipe()
        .name(pipe_name)
        .source(sqs_arn(src_queue))
        .target(sqs_arn(tgt_queue))
        .role_arn(ROLE_ARN)
        .desired_state(aws_sdk_pipes::types::RequestedPipeState::Running)
        .source_parameters(
            aws_sdk_pipes::types::PipeSourceParameters::builder()
                .filter_criteria(
                    aws_sdk_pipes::types::FilterCriteria::builder()
                        .filters(
                            aws_sdk_pipes::types::Filter::builder()
                                .pattern(r#"{"body": {"status": ["active"]}}"#)
                                .build(),
                        )
                        .build(),
                )
                .build(),
        )
        .send()
        .await
        .expect("create pipe");

    sqs.send_message()
        .queue_url(&src_url)
        .message_body(r#"{"status": "active", "id": "match-1"}"#)
        .send()
        .await
        .expect("send matching message");

    sqs.send_message()
        .queue_url(&src_url)
        .message_body(r#"{"status": "inactive", "id": "no-match"}"#)
        .send()
        .await
        .expect("send non-matching message");

    let mut found = false;
    for _ in 0..15 {
        tokio::time::sleep(std::time::Duration::from_secs(1)).await;
        let recv = sqs
            .receive_message()
            .queue_url(&tgt_url)
            .max_number_of_messages(10)
            .wait_time_seconds(1)
            .send()
            .await;
        if let Ok(r) = recv {
            if let Some(msgs) = r.messages() {
                if msgs.iter().any(|m| m.body().unwrap_or("").contains("match-1")) {
                    assert!(
                        !msgs.iter().any(|m| m.body().unwrap_or("").contains("no-match")),
                        "non-matching message should not be forwarded"
                    );
                    found = true;
                    break;
                }
            }
        }
    }
    assert!(found, "target queue should receive matching message");

    let attrs = sqs
        .get_queue_attributes()
        .queue_url(&src_url)
        .attribute_names(aws_sdk_sqs::types::QueueAttributeName::ApproximateNumberOfMessages)
        .send()
        .await
        .expect("get queue attributes");
    assert_eq!(
        attrs.attributes().unwrap().get(&aws_sdk_sqs::types::QueueAttributeName::ApproximateNumberOfMessages).map(|s| s.as_str()),
        Some("0"),
        "source queue should be drained"
    );
}

#[tokio::test(flavor = "multi_thread")]
async fn test_pipes_batch_size() {
    let pipes = common::pipes_client().await;
    let sqs = common::sqs_client().await;
    let src_queue = "rust-pipe-src-batch";
    let tgt_queue = "rust-pipe-tgt-batch";
    let pipe_name = "rust-pipe-batch";

    let src_resp = sqs.create_queue().queue_name(src_queue).send().await.expect("create src queue");
    let src_url = src_resp.queue_url().unwrap_or("").to_string();
    let tgt_resp = sqs.create_queue().queue_name(tgt_queue).send().await.expect("create tgt queue");
    let tgt_url = tgt_resp.queue_url().unwrap_or("").to_string();

    let _guard = common::CleanupGuard::new({
        let pipes = pipes.clone();
        let sqs = sqs.clone();
        let src_url = src_url.clone();
        let tgt_url = tgt_url.clone();
        async move {
            let _ = pipes.delete_pipe().name(pipe_name).send().await;
            let _ = sqs.delete_queue().queue_url(&src_url).send().await;
            let _ = sqs.delete_queue().queue_url(&tgt_url).send().await;
        }
    });

    pipes
        .create_pipe()
        .name(pipe_name)
        .source(sqs_arn(src_queue))
        .target(sqs_arn(tgt_queue))
        .role_arn(ROLE_ARN)
        .desired_state(aws_sdk_pipes::types::RequestedPipeState::Running)
        .source_parameters(
            aws_sdk_pipes::types::PipeSourceParameters::builder()
                .sqs_queue_parameters(
                    aws_sdk_pipes::types::PipeSourceSqsQueueParameters::builder()
                        .batch_size(1)
                        .build(),
                )
                .build(),
        )
        .send()
        .await
        .expect("create pipe");

    for i in 1..=3 {
        sqs.send_message()
            .queue_url(&src_url)
            .message_body(format!("batch-msg-{}", i))
            .send()
            .await
            .expect("send message");
    }

    let mut found_messages = std::collections::HashSet::new();
    for _ in 0..20 {
        if found_messages.len() >= 3 {
            break;
        }
        tokio::time::sleep(std::time::Duration::from_secs(1)).await;
        let recv = sqs
            .receive_message()
            .queue_url(&tgt_url)
            .max_number_of_messages(10)
            .wait_time_seconds(1)
            .send()
            .await;
        if let Ok(r) = recv {
            if let Some(msgs) = r.messages() {
                for msg in msgs {
                    for j in 1..=3 {
                        let key = format!("batch-msg-{}", j);
                        if msg.body().unwrap_or("").contains(&key) {
                            found_messages.insert(key);
                        }
                    }
                }
            }
        }
    }
    assert_eq!(found_messages.len(), 3, "all 3 messages should arrive at target");
}

#[tokio::test(flavor = "multi_thread")]
async fn test_pipes_stopped_pipe_does_not_forward() {
    let pipes = common::pipes_client().await;
    let sqs = common::sqs_client().await;
    let src_queue = "rust-pipe-src-nofwd";
    let tgt_queue = "rust-pipe-tgt-nofwd";
    let pipe_name = "rust-pipe-nofwd";

    let src_resp = sqs.create_queue().queue_name(src_queue).send().await.expect("create src queue");
    let src_url = src_resp.queue_url().unwrap_or("").to_string();
    let tgt_resp = sqs.create_queue().queue_name(tgt_queue).send().await.expect("create tgt queue");
    let tgt_url = tgt_resp.queue_url().unwrap_or("").to_string();

    let _guard = common::CleanupGuard::new({
        let pipes = pipes.clone();
        let sqs = sqs.clone();
        let src_url = src_url.clone();
        let tgt_url = tgt_url.clone();
        async move {
            let _ = pipes.delete_pipe().name(pipe_name).send().await;
            let _ = sqs.delete_queue().queue_url(&src_url).send().await;
            let _ = sqs.delete_queue().queue_url(&tgt_url).send().await;
        }
    });

    pipes
        .create_pipe()
        .name(pipe_name)
        .source(sqs_arn(src_queue))
        .target(sqs_arn(tgt_queue))
        .role_arn(ROLE_ARN)
        .desired_state(aws_sdk_pipes::types::RequestedPipeState::Stopped)
        .send()
        .await
        .expect("create pipe");

    sqs.send_message()
        .queue_url(&src_url)
        .message_body("should not forward")
        .send()
        .await
        .expect("send message");

    tokio::time::sleep(std::time::Duration::from_secs(3)).await;

    let recv = sqs
        .receive_message()
        .queue_url(&tgt_url)
        .max_number_of_messages(1)
        .wait_time_seconds(1)
        .send()
        .await
        .expect("receive message");
    assert!(
        recv.messages().is_empty(),
        "target queue should be empty"
    );
}

async fn cleanup_queue(sqs: &aws_sdk_sqs::Client, queue_name: &str) {
    if let Ok(r) = sqs.get_queue_url().queue_name(queue_name).send().await {
        if let Some(url) = r.queue_url() {
            let _ = sqs.delete_queue().queue_url(url).send().await;
        }
    }
}
